### PR TITLE
Add "rust-phf: the perfect hash function" blog post

### PR DESCRIPTION
**Pull request by Turbo87**
[Rendered](https://deploy-preview-1788--objective-northcutt-37494c.netlify.app/blog/2022/04/26/the-perfect-hash-function/)

### CODE REVIEW

File: _posts/2022-04-26-the-perfect-hash-function.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+If you have read-only mappings that you want to use in your Rust code, then the
+`rust-phf` project can give you significant speed improvements by moving some of
+the work to compile time.
+
+We hope that this short story about our work on `conduit-mime-types` was helpful
+to you. If you have any questions do not hesitate to [contact] us. We're happy
+to help!
+
+[contact]: https://simplabs.com/contact/
```
Comment (oscard0m) [inline _posts/2022-04-26-the-perfect-hash-function.md]: 💅🏽 The links for `[crates.io]` and `[Rust]` are missing

```suggestion
[crates.io]: https://crates.io/
[Rust]: https://www.rust-lang.org/
[contact]: https://simplabs.com/contact/
```

**Review by oscard0m — APPROVED** [review]

File: _posts/2022-04-26-the-perfect-hash-function.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+This is the story of how we made the [conduit-mime-types] Rust crate almost
+infinitely faster, using perfect hash functions and compile-time code
+generation.
+
+[conduit-mime-types]: https://github.com/conduit-rust/conduit-mime-types
+
+<!--break-->
+
+Let's start at the beginning. [crates.io] is the package registry of the [Rust]
```
Comment (marcoow) [inline _posts/2022-04-26-the-perfect-hash-function.md]: These links don't work properly:

Bildschirmfoto 2022-04-26 um 08 01 01

File: _posts/2022-04-26-the-perfect-hash-function.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+improve the initialization speed.
+
+## Step 1: Code generation
+
+While JSON can be parsed at
+[blazing speed](https://github.com/simdjson/simdjson) these days, it is still
+slower than not having to parse it at all. Our first step towards a more
+efficient implementation was to get rid of the JSON parsing step.
+
+How did we get rid of the parsing step? Well, we didn't, but we did moved it
```
Comment (marcoow) [inline _posts/2022-04-26-the-perfect-hash-function.md]:
```suggestion
How did we get rid of the parsing step? Well, we didn't, but we did move it
```

Comment (marcoow) [reply _posts/2022-04-26-the-perfect-hash-function.md]: …or

```suggestion
How did we get rid of the parsing step? Well, we didn't, but we moved it
```

of course :)

**Review by marcoow — APPROVED** [review]
Nice! Also well written and easy to follow.

Just some minor points to fix…

File: _posts/2022-04-26-the-perfect-hash-function.md (hunk 3 above)
Comment (Turbo87) [reply _posts/2022-04-26-the-perfect-hash-function.md]: fixed, thanks :)

File: _posts/2022-04-26-the-perfect-hash-function.md (hunk 1 above)
Comment (Turbo87) [reply _posts/2022-04-26-the-perfect-hash-function.md]: fixed, thanks :)

File: _posts/2022-04-26-the-perfect-hash-function.md (hunk 2 above)
Comment (Turbo87) [reply _posts/2022-04-26-the-perfect-hash-function.md]: see https://github.com/simplabs/simplabs.github.io/pull/1788#discussion_r857886707 :)

File: _posts/2022-04-26-the-perfect-hash-function.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+The `main()` function of this file will be executed automatically before your
+library is compiled, and you can use it to generate arbitrary code files that
+the rest of your code can then import, or rather `include!(...)`.
+
+While this admittedly decreased our build speed, it also meant that the raw JSON
+data and the generated code file could no longer diverge and cause subtle bugs.
+In practice the build speed degradation was barely measurable though, since the
+compilation of the code itself already took a significant amount of time.
+
+### Step 3: Perfect hash functions
```
Comment (marcoow) [inline _posts/2022-04-26-the-perfect-hash-function.md]:
```suggestion
## Step 3: Perfect hash functions
```

Just recognized the others use `<h2>` – bit of a nitpick but once you see it it's strange :)

Comment (Turbo87) [reply _posts/2022-04-26-the-perfect-hash-function.md]: yeah, definitely unintended 😅

**Comment by marcoow** [comment]
@emmasofiah2o when is this going out? It was written more than a month ago…

**Comment by Turbo87** [comment]
@emmasofiah2o @marcoow it's been almost two months now. how do we want to proceed with this? should I just merge and promote it myself?

**Comment by emmasofiah2o** [comment]
@Turbo87 I thought you said this could be posted as your rotation blog post... We are still very behind on the rotation and I keep having to postpone which is why the release got delayed so much. I am on vacation next week and I don't know if I can get a post ready for this to be published before then. I can schedule it for the 12th of July and switch the release date of Ricardo's post if that works. If it's super urgent I can try to prepare the post and ask Sarah to release it, but you'd have to do the merging and date changing yourself.

**Comment by emmasofiah2o** [comment]
I believe you also still need an og:image for this?

**Comment by Turbo87** [comment]
> I thought you said this could be posted as your rotation blog post

yeah, I guess that's fine too.

> I believe you also still need an og:image for this?

I don't know. if we require this then someone will have to create one, yes.

**Comment by emmasofiah2o** [comment]
> yeah, I guess that's fine too.
If I use it for your rotation it will go out in September realistically. I'd suggest I post it on the 12th instead of Ricardo's, give him a bit more time to work on his post and then have you skip your turn on the rotation. That fine with you?

> I don't know. if we require this then someone will have to create one, yes.
Yes, every blog post has one. I'll take care of it for you when I release, it's not a lot of work :)

**Comment by marcoow** [comment]
We shouldn't delay posts when there are posts way before the person's due in the rotation – the rotation is the minimum and if people write more that's great :)

@emmasofiah2o can you add an `og:image` and release this?

**Comment by emmasofiah2o** [comment]
> We shouldn't delay posts when there are posts way before the person's due in the rotation – the rotation is the minimum and if people write more that's great :)

Got it. I asked at the time and was told to use it for the rotation so I figured it wouldn't need to be released right away. I'll use this as a rule of thumb in case this happens again then. 

> @emmasofiah2o can you add an `og:image` and release this?

Do you want it out this week? We have quite the full schedule already with all the Ember Europe posts. I could get it out on Friday, but it would get a bit less visibility.

File: _posts/2022-04-26-the-perfect-hash-function.md [hunk 5]
```diff
@@ -5,6 +5,9 @@ github: Turbo87
 twitter: tobiasbieniek
 topic: rust
 bio: 'Senior Software Engineer'
+description: 'Insert description'
```
Comment (emmasofiah2o) [inline _posts/2022-04-26-the-perfect-hash-function.md]: @Turbo87 can you add a short description please? I noticed it was missing :)

**Comment by marcoow** [comment]
Ember Europe is unrelated so yes, I'd say let's release this as soon as we can and not delay longer

File: _posts/2022-04-26-the-perfect-hash-function.md (hunk 5 above)
Comment (Turbo87) [reply _posts/2022-04-26-the-perfect-hash-function.md]: what is that supposed to describe and where is it being used? is it the same as what is being posted to twitter etc.?

Comment (marcoow) [reply _posts/2022-04-26-the-perfect-hash-function.md]: yeah, kind of – that's what we're also using in metatags and so on, basically a 1 sentence summary

**Pull request merged**